### PR TITLE
Networking: Add libkano_networking library

### PR DESCRIPTION
**Pull request by tombettany**
Provide optimised networking functions implemented in C with a Python
interface.

@convolu @Ealdwulf to review.

### CODE REVIEW

File: libs/kano-networking/src/ifaces.c [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    freeifaddrs(iface_addr);
+
+    return 0;
+}
+
+/**
+ * Check for interfaces of the provided type and print the interface name
+ */
+int check_iface_type(char * iface_type)
```
Comment (convolu) [inline libs/kano-networking/src/ifaces.c]: Are you altering the iface_type string anywhere? Could it be a `const char *` for clarity?

File: libs/kano-networking/src/ifaces.c [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Search through the available interfaces and return the first available
+ * which matches the type provided
+ *
+ * The caller is responsible for freeing iface_name
+ *
+ * Upon error, iface_name does not need freeing
+ */
+int select_iface(char * iface_type, char **iface_name)
```
Comment (convolu) [inline libs/kano-networking/src/ifaces.c]: Do you think that the `iface_type` can be a `const char *` ?

File: libs/kano-networking/includes/ifaces.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+#ifndef __KANO_NETWORKING_IFACES__
+#define __KANO_NETWORKING_IFACES__
+
+enum ERRORS {
+    E_NO_INTERFACES = -1
+};
+
+int select_iface(char * iface_type, char **iface_name);
+int check_iface_type(char * iface_type);
```
Comment (convolu) [inline libs/kano-networking/includes/ifaces.h]: Personally I like my typedefs, so I would typedef the enum and then have that be the return value of these two fns, but it is clear as it is anyway

**Comment by tombettany** [comment]
Won't merge until release of v2.3.0

File: libs/kano-networking/src/ifaces.c [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        candidate_iface = candidate_iface->ifa_next;
+    }
+
+    if (!match) {
+        freeifaddrs(iface_addr);
+        return E_NO_INTERFACES;
+    }
+
+    // Allow for termination character
+    size_t len = strlen(candidate_iface->ifa_name) + 1;
```
Comment (Ealdwulf) [inline libs/kano-networking/src/ifaces.c]: I think these three lines are the same as  `*iface_name = strdup(candidate_iface->ifa_name);` ?

**Comment by Ealdwulf** [comment]
Apart from minor point above, looks good :+1:

File: debian/control [hunk 5]
```diff
@@ -6,10 +6,11 @@ Standards-Version: 3.9.2
 Build-Depends: debhelper (>= 9), libraspberrypi-dev, libpng12-dev
 
 Package: kano-toolset
-Architecture: all
+Architecture: any
```
Comment (radujipa) [inline debian/control]: Is there a difference between the two here? Most of our packages use `all`.

Comment (tombettany) [inline debian/control]: `all` means that the built package can be installed on any architecture without any changes (e.g. packages only containing python scripts, shell scripts, things which do not need compilation). `any` means that the package can be installed on any architecture but needs to be rebuilt for the platform for this to be possible.

Comment (convolu) [inline debian/control]: Thanks for clearing this up @tombettany I was confused as well

File: libs/kano-networking/src/ifaces.c (hunk 4 above)
Comment (tombettany) [inline libs/kano-networking/src/ifaces.c]: Good spot, you are indeed right.

File: debian/control (hunk 5 above)
Comment (radujipa) [inline debian/control]: Ah right :+1: Thanks for the explanation :)

**Comment by tombettany** [comment]
Merging now that we have released.

**Pull request merged**